### PR TITLE
Privatize iexamine function in map_data_common_t

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
SUMMARY: None

#### Purpose of change
Reduce ugly function pointer comparisons, and pave the way for future changes.
There should be no changes in function.

#### Describe the solution
Privatize the examine function, add functions to check what type of function is is, and a function to forward to it.

#### Testing
Compiles. I've looked over my changes several times to make sure it doesn't change anything functionally.

### CODE REVIEW

**Review by Aivean — COMMENTED** [review]
I'm not sure that adding a bunch of methods for each `iexamine::*` function is a good idea.
Please consider an alternative:

```cpp
using iexamine_function_ref = void ( & )( player &, const tripoint & );

struct map_data_common_t {
    bool has_examine( iexamine_function_ref f ) const {
        return examine == &f;
    }
}


TEST_CASE( "mapdata_examine" )
{
    map_data_common_t t;
    t.examine = iexamine::water_source;


    CHECK( t.has_examine( iexamine::water_source ) );
    CHECK_FALSE( t.has_examine( iexamine::fungus ) );
    CHECK_FALSE( t.has_examine( iexamine::dirtmound ) );
    CHECK_FALSE( t.has_examine( iexamine::none ) );

}
```

**Comment by anothersimulacrum** [comment]
Good idea.

**Review by Aivean — APPROVED** [review]

File: src/gates.cpp [hunk 1]
```diff
@@ -102,12 +102,11 @@ void gate_data::load( const JsonObject &jo, const std::string & )
 
 void gate_data::check() const
 {
-    static const iexamine_function controls_gate( iexamine_function_from_string( "controls_gate" ) );
     const ter_str_id winch_tid( id.str() );
 
     if( !winch_tid.is_valid() ) {
         debugmsg( "Gates \"%s\" have no terrain of the same name, working as a winch.", id.c_str() );
-    } else if( winch_tid->examine != controls_gate ) {
+    } else if( !winch_tid->has_examine( *iexamine::controls_gate ) ) {
```
Comment (kevingranade) [inline src/gates.cpp:109]:
```suggestion
    } else if( !winch_tid->has_examine( iexamine::controls_gate ) ) {
```

**Comment by kevingranade** [comment]
The pointer comparison isn't *bad* per se, but it's a bit low level to have scattered all over the place, so I agree hiding it like this is an improvement.

Also half way through reading this "examine" stopped looking like a word, so I was like, "wait how is this compiling if it's spelled 'ex-amine'?"

File: src/gates.cpp (hunk 1 above)
Comment (anothersimulacrum) [reply src/gates.cpp]: Ah, whoops, thanks.

**Pull request merged**